File: reference/python/test_var.py
"""Tests for the VaR reference implementation."""

from __future__ import annotations

import os

import numpy as np
import pandas as pd
import pytest

from var import (
    log_returns,
    var_historical,
    cvar_historical,
    var_parametric,
    cvar_parametric,
    simulate_gbm_paths,
    var_monte_carlo,
    kupiec_test,
    backtest_var,
    compute_var_surface,
)

FIXTURES = os.path.join(os.path.dirname(__file__), "..", "fixtures")


# ---------------------------------------------------------------------------
# Fixture loading
# ---------------------------------------------------------------------------

@pytest.fixture
def sample_returns() -> np.ndarray:
    df = pd.read_csv(os.path.join(FIXTURES, "sample_returns.csv"))
    return df["return"].values


# ---------------------------------------------------------------------------
# log_returns
# ---------------------------------------------------------------------------

class TestLogReturns:
    def test_length(self):
        prices = pd.Series([100.0, 105.0, 103.0, 107.0])
        lr = log_returns(prices)
        assert len(lr) == len(prices) - 1

    def test_values(self):
        prices = pd.Series([100.0, 110.0])
        lr = log_returns(prices)
        assert lr.iloc[0] == pytest.approx(np.log(1.1))

    def test_round_trip(self):
        prices = pd.Series([100.0, 105.0, 103.0, 107.0, 102.0])
        lr = log_returns(prices)
        reconstructed = prices.iloc[0] * np.exp(lr.cumsum())
        np.testing.assert_allclose(reconstructed.values, prices.iloc[1:].values)


# ---------------------------------------------------------------------------
# Historical Simulation
# ---------------------------------------------------------------------------

class TestHistoricalVar:
    def test_positive(self, sample_returns):
        assert var_historical(sample_returns) > 0

    def test_higher_confidence_higher_var(self, sample_returns):
        v90 = var_historical(sample_returns, confidence=0.90)
        v95 = var_historical(sample_returns, confidence=0.95)
        v99 = var_historical(sample_returns, confidence=0.99)
        assert v90 < v95 < v99

    def test_longer_horizon_higher_var(self, sample_returns):
        v1 = var_historical(sample_returns, holding_period=1)
        v5 = var_historical(sample_returns, holding_period=5)
        v10 = var_historical(sample_returns, holding_period=10)
        assert v1 < v5 < v10

    def test_sqrt_scaling(self, sample_returns):
        v1 = var_historical(sample_returns, holding_period=1)
        v5 = var_historical(sample_returns, holding_period=5)
        assert v5 == pytest.approx(v1 * np.sqrt(5))


class TestHistoricalCvar:
    def test_positive(self, sample_returns):
        assert cvar_historical(sample_returns) > 0

    def test_cvar_geq_var(self, sample_returns):
        for conf in [0.90, 0.95, 0.99]:
            v = var_historical(sample_returns, confidence=conf)
            cv = cvar_historical(sample_returns, confidence=conf)
            assert cv >= v, f"CVaR ({cv}) < VaR ({v}) at {conf}"


# ---------------------------------------------------------------------------
# Parametric (Variance-Covariance)
# ---------------------------------------------------------------------------

class TestParametricVar:
    def test_positive(self, sample_returns):
        assert var_parametric(sample_returns) > 0

    def test_higher_confidence_higher_var(self, sample_returns):
        v90 = var_parametric(sample_returns, confidence=0.90)
        v95 = var_parametric(sample_returns, confidence=0.95)
        v99 = var_parametric(sample_returns, confidence=0.99)
        assert v90 < v95 < v99

    def test_longer_horizon_higher_var(self, sample_returns):
        v1 = var_parametric(sample_returns, holding_period=1)
        v5 = var_parametric(sample_returns, holding_period=5)
        v10 = var_parametric(sample_returns, holding_period=10)
        assert v1 < v5 < v10

    def test_known_normal(self):
        """Sanity check: for zero-mean normal data, VaR should be z * sigma."""
        rng = np.random.default_rng(99)
        r = rng.normal(0, 0.02, 100_000)
        v = var_parametric(r, confidence=0.95, holding_period=1)
        expected = 1.6449 * 0.02
        assert v == pytest.approx(expected, rel=0.02)


class TestParametricCvar:
    def test_positive(self, sample_returns):
        assert cvar_parametric(sample_returns) > 0

    def test_cvar_geq_var(self, sample_returns):
        for conf in [0.90, 0.95, 0.99]:
            v = var_parametric(sample_returns, confidence=conf)
            cv = cvar_parametric(sample_returns, confidence=conf)
            assert cv >= v, f"CVaR ({cv}) < VaR ({v}) at {conf}"


# ---------------------------------------------------------------------------
# Monte Carlo Simulation
# ---------------------------------------------------------------------------

class TestSimulateGbmPaths:
    def test_shape(self):
        paths = simulate_gbm_paths(S0=100.0, mu=0.0, sigma=0.2, holding_period=10, n_simulations=500)
        assert paths.shape == (500, 11)

    def test_starts_at_s0(self):
        paths = simulate_gbm_paths(S0=42.0, mu=0.05, sigma=0.15, holding_period=5, n_simulations=100)
        np.testing.assert_allclose(paths[:, 0], 42.0)

    def test_seed_reproducibility(self):
        p1 = simulate_gbm_paths(S0=100.0, mu=0.0, sigma=0.2, holding_period=10, seed=7)
        p2 = simulate_gbm_paths(S0=100.0, mu=0.0, sigma=0.2, holding_period=10, seed=7)
        np.testing.assert_array_equal(p1, p2)

    def test_all_positive(self):
        paths = simulate_gbm_paths(S0=100.0, mu=-0.1, sigma=0.5, holding_period=20, n_simulations=1000)
        assert np.all(paths > 0)

    def test_custom_steps(self):
        paths = simulate_gbm_paths(S0=100.0, mu=0.0, sigma=0.2, holding_period=5, n_simulations=100, n_steps=25)
        assert paths.shape == (100, 26)


class TestMonteCarloVar:
    def test_positive(self, sample_returns):
        result = var_monte_carlo(sample_returns)
        assert result["var"] > 0

    def test_cvar_geq_var(self, sample_returns):
        result = var_monte_carlo(sample_returns, confidence=0.95)
        assert result["cvar"] >= result["var"]

    def test_seed_reproducibility(self, sample_returns):
        r1 = var_monte_carlo(sample_returns, seed=123)
        r2 = var_monte_carlo(sample_returns, seed=123)
        assert r1["var"] == r2["var"]
        assert r1["cvar"] == r2["cvar"]

    def test_simulated_returns_length(self, sample_returns):
        n = 5_000
        result = var_monte_carlo(sample_returns, n_simulations=n)
        assert len(result["simulated_returns"]) == n

    def test_converges_to_parametric(self):
        """MC should land close to parametric when the data is actually normal."""
        rng = np.random.default_rng(99)
        r = rng.normal(0, 0.02, 10_000)
        mc = var_monte_carlo(r, confidence=0.95, n_simulations=200_000, seed=42)
        param = var_parametric(r, confidence=0.95)
        assert mc["var"] == pytest.approx(param, rel=0.05)

    def test_higher_confidence_higher_var(self, sample_returns):
        v90 = var_monte_carlo(sample_returns, confidence=0.90)["var"]
        v95 = var_monte_carlo(sample_returns, confidence=0.95)["var"]
        v99 = var_monte_carlo(sample_returns, confidence=0.99)["var"]
        assert v90 < v95 < v99


# ---------------------------------------------------------------------------
# Backtesting
# ---------------------------------------------------------------------------

class TestKupiecTest:
    def test_well_calibrated(self):
        """50 breaches out of 1000 at 95% - should not reject."""
        p_value = kupiec_test(n_observations=1000, n_breaches=50, confidence=0.95)
        assert p_value > 0.05

    def test_too_many_breaches(self):
        """120 breaches out of 1000 at 95% - clearly miscalibrated."""
        p_value = kupiec_test(n_observations=1000, n_breaches=120, confidence=0.95)
        assert p_value < 0.05

    def test_edge_zero_breaches(self):
        p_value = kupiec_test(n_observations=1000, n_breaches=0, confidence=0.95)
        assert p_value == 1.0

    def test_edge_all_breaches(self):
        p_value = kupiec_test(n_observations=100, n_breaches=100, confidence=0.95)
        assert p_value == 1.0


class TestBacktestVar:
    def test_output_shape(self, sample_returns):
        df = backtest_var(sample_returns, window=250)
        assert len(df) == len(sample_returns) - 250
        assert set(df.columns) == {"date_index", "actual_return", "var_prediction", "breach"}

    def test_breach_count_within_bounds(self, sample_returns):
        df = backtest_var(sample_returns, confidence=0.95, window=250)
        assert 0 <= df.attrs["breach_count"] <= len(df)

    def test_breach_rate_reasonable(self, sample_returns):
        """Synthetic data is well-behaved, so breach rate shouldn't be wild."""
        df = backtest_var(sample_returns, confidence=0.95, window=250)
        assert 0.0 < df.attrs["breach_rate"] < 0.20

    def test_expected_breach_rate(self, sample_returns):
        df = backtest_var(sample_returns, confidence=0.95)
        assert df.attrs["expected_breach_rate"] == pytest.approx(0.05)

    def test_kupiec_attached(self, sample_returns):
        df = backtest_var(sample_returns, confidence=0.95)
        assert 0.0 <= df.attrs["kupiec_p_value"] <= 1.0

    def test_historical_method(self, sample_returns):
        df = backtest_var(sample_returns, method="historical")
        assert len(df) > 0

    def test_parametric_method(self, sample_returns):
        df = backtest_var(sample_returns, method="parametric")
        assert len(df) > 0

    def test_invalid_method(self, sample_returns):
        with pytest.raises(ValueError, match="Unknown method"):
            backtest_var(sample_returns, method="invalid")


# ---------------------------------------------------------------------------
# VaR Surface
# ---------------------------------------------------------------------------

class TestVarSurface:
    def test_shape(self, sample_returns):
        df = compute_var_surface(sample_returns)
        assert len(df) == 9  # 3 confidences x 3 holding periods

    def test_columns(self, sample_returns):
        df = compute_var_surface(sample_returns)
        expected_cols = {
            "confidence", "holding_period",
            "var_historical", "cvar_historical",
            "var_parametric", "cvar_parametric",
            "var_monte_carlo", "cvar_monte_carlo",
        }
        assert set(df.columns) == expected_cols

    def test_all_positive(self, sample_returns):
        df = compute_var_surface(sample_returns)
        value_cols = [c for c in df.columns if c.startswith("var_") or c.startswith("cvar_")]
        for col in value_cols:
            assert (df[col] > 0).all(), f"{col} contains non-positive values"

    def test_cvar_geq_var(self, sample_returns):
        df = compute_var_surface(sample_returns)
        for method in ["historical", "parametric", "monte_carlo"]:
            assert (df[f"cvar_{method}"] >= df[f"var_{method}"]).all()

    def test_monotonic_in_confidence(self, sample_returns):
        df = compute_var_surface(sample_returns)
        for hp in [1, 5, 10]:
            subset = df[df["holding_period"] == hp].sort_values("confidence")
            for method in ["historical", "parametric", "monte_carlo"]:
                vals = subset[f"var_{method}"].values
                assert all(vals[i] < vals[i + 1] for i in range(len(vals) - 1))

    def test_monotonic_in_holding_period(self, sample_returns):
        df = compute_var_surface(sample_returns)
        for conf in [0.90, 0.95, 0.99]:
            subset = df[df["confidence"] == conf].sort_values("holding_period")
            for method in ["historical", "parametric", "monte_carlo"]:
                vals = subset[f"var_{method}"].values
                assert all(vals[i] < vals[i + 1] for i in range(len(vals) - 1))

    def test_custom_params(self, sample_returns):
        df = compute_var_surface(sample_returns, confidences=[0.95], holding_periods=[1, 5])
        assert len(df) == 2


# ---------------------------------------------------------------------------
# Fixture validation
# ---------------------------------------------------------------------------

class TestFixtureValidation:
    """Check we match the locked-down fixture values exactly."""

    def test_historical_fixture(self, sample_returns):
        expected = pd.read_csv(os.path.join(FIXTURES, "expected_historical.csv"))
        for _, row in expected.iterrows():
            v = var_historical(sample_returns, row["confidence"], int(row["holding_period"]))
            cv = cvar_historical(sample_returns, row["confidence"], int(row["holding_period"]))
            assert v == pytest.approx(row["var"], rel=1e-6), \
                f"Historical VaR mismatch at conf={row['confidence']}, hp={row['holding_period']}"
            assert cv == pytest.approx(row["cvar"], rel=1e-6), \
                f"Historical CVaR mismatch at conf={row['confidence']}, hp={row['holding_period']}"

    def test_parametric_fixture(self, sample_returns):
        expected = pd.read_csv(os.path.join(FIXTURES, "expected_parametric.csv"))
        for _, row in expected.iterrows():
            v = var_parametric(sample_returns, row["confidence"], int(row["holding_period"]))
            cv = cvar_parametric(sample_returns, row["confidence"], int(row["holding_period"]))
            assert v == pytest.approx(row["var"], rel=1e-6), \
                f"Parametric VaR mismatch at conf={row['confidence']}, hp={row['holding_period']}"
            assert cv == pytest.approx(row["cvar"], rel=1e-6), \
                f"Parametric CVaR mismatch at conf={row['confidence']}, hp={row['holding_period']}"

    def test_monte_carlo_fixture(self, sample_returns):
        expected = pd.read_csv(os.path.join(FIXTURES, "expected_monte_carlo.csv"))
        for _, row in expected.iterrows():
            mc = var_monte_carlo(
                sample_returns, row["confidence"], int(row["holding_period"]),
                n_simulations=100_000, seed=42,
            )
            assert mc["var"] == pytest.approx(row["var"], rel=1e-6), \
                f"MC VaR mismatch at conf={row['confidence']}, hp={row['holding_period']}"
            assert mc["cvar"] == pytest.approx(row["cvar"], rel=1e-6), \
                f"MC CVaR mismatch at conf={row['confidence']}, hp={row['holding_period']}"

File: reference/cpp/test_var.cpp
#include "var.hpp"

#include <cassert>
#include <cmath>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

namespace {

// Bare-bones CSV reader - just enough for our fixture format.
std::vector<std::vector<std::string>> read_csv(const std::string& path) {
    std::ifstream file(path);
    if (!file) {
        throw std::runtime_error("Could not open fixture file: " + path);
    }
    std::vector<std::vector<std::string>> rows;
    std::string line;
    bool first = true;
    while (std::getline(file, line)) {
        if (first) { first = false; continue; }
        if (line.empty()) continue;
        std::vector<std::string> fields;
        std::stringstream ss(line);
        std::string field;
        while (std::getline(ss, field, ',')) {
            fields.push_back(field);
        }
        rows.push_back(fields);
    }
    return rows;
}

std::vector<double> load_returns(const std::string& path) {
    std::vector<double> returns;
    for (const auto& row : read_csv(path)) {
        returns.push_back(std::stod(row[0]));
    }
    return returns;
}

struct ExpectedRow {
    double confidence;
    int holding_period;
    double var;
    double cvar;
};

std::vector<ExpectedRow> load_expected(const std::string& path) {
    std::vector<ExpectedRow> rows;
    for (const auto& row : read_csv(path)) {
        rows.push_back({std::stod(row[0]), std::stoi(row[1]),
                         std::stod(row[2]), std::stod(row[3])});
    }
    return rows;
}

void assert_approx(double actual, double expected, double rel_tol,
                   const std::string& label) {
    double diff = std::abs(actual - expected);
    double threshold = rel_tol * std::abs(expected);
    if (diff > threshold && diff > 1e-10) {
        std::cerr << "FAIL: " << label
                  << " - expected " << expected
                  << ", got " << actual
                  << " (diff " << diff << ")" << std::endl;
        assert(false);
    }
}

std::string fixtures_dir() {
#ifdef FIXTURES_DIR
    return FIXTURES_DIR;
#else
    return "../fixtures";
#endif
}

int tests_run = 0;
int tests_passed = 0;

void pass(const std::string& name) {
    ++tests_run;
    ++tests_passed;
    std::cout << "  PASS  " << name << std::endl;
}

}  // namespace

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

void test_log_returns() {
    std::vector<double> prices = {100.0, 105.0, 103.0, 107.0};
    auto lr = var::log_returns(prices);
    assert(lr.size() == 3);
    assert_approx(lr[0], std::log(105.0 / 100.0), 1e-10, "log_returns[0]");
    pass("log_returns");
}

void test_historical_positive(const std::vector<double>& returns) {
    assert(var::var_historical(returns) > 0);
    pass("historical_var_positive");
}

void test_historical_monotonic_confidence(const std::vector<double>& returns) {
    double v90 = var::var_historical(returns, 0.90);
    double v95 = var::var_historical(returns, 0.95);
    double v99 = var::var_historical(returns, 0.99);
    assert(v90 < v95 && v95 < v99);
    pass("historical_var_monotonic_confidence");
}

void test_historical_monotonic_horizon(const std::vector<double>& returns) {
    double v1 = var::var_historical(returns, 0.95, 1);
    double v5 = var::var_historical(returns, 0.95, 5);
    double v10 = var::var_historical(returns, 0.95, 10);
    assert(v1 < v5 && v5 < v10);
    pass("historical_var_monotonic_horizon");
}

void test_historical_cvar_geq_var(const std::vector<double>& returns) {
    for (double conf : {0.90, 0.95, 0.99}) {
        double v = var::var_historical(returns, conf);
        double cv = var::cvar_historical(returns, conf);
        assert(cv >= v);
    }
    pass("historical_cvar_geq_var");
}

void test_parametric_positive(const std::vector<double>& returns) {
    assert(var::var_parametric(returns) > 0);
    pass("parametric_var_positive");
}

void test_parametric_monotonic_confidence(const std::vector<double>& returns) {
    double v90 = var::var_parametric(returns, 0.90);
    double v95 = var::var_parametric(returns, 0.95);
    double v99 = var::var_parametric(returns, 0.99);
    assert(v90 < v95 && v95 < v99);
    pass("parametric_var_monotonic_confidence");
}

void test_parametric_cvar_geq_var(const std::vector<double>& returns) {
    for (double conf : {0.90, 0.95, 0.99}) {
        double v = var::var_parametric(returns, conf);
        double cv = var::cvar_parametric(returns, conf);
        assert(cv >= v);
    }
    pass("parametric_cvar_geq_var");
}

void test_monte_carlo_positive(const std::vector<double>& returns) {
    auto result = var::var_monte_carlo(returns);
    assert(result.var > 0);
    pass("monte_carlo_var_positive");
}

void test_monte_carlo_cvar_geq_var(const std::vector<double>& returns) {
    auto result = var::var_monte_carlo(returns, 0.95);
    assert(result.cvar >= result.var);
    pass("monte_carlo_cvar_geq_var");
}

void test_monte_carlo_seed_reproducibility(const std::vector<double>& returns) {
    auto r1 = var::var_monte_carlo(returns, 0.95, 1, 10000, 123);
    auto r2 = var::var_monte_carlo(returns, 0.95, 1, 10000, 123);
    assert(r1.var == r2.var);
    assert(r1.cvar == r2.cvar);
    pass("monte_carlo_seed_reproducibility");
}

void test_monte_carlo_simulated_returns_length(const std::vector<double>& returns) {
    auto result = var::var_monte_carlo(returns, 0.95, 1, 5000);
    assert(result.simulated_returns.size() == 5000);
    pass("monte_carlo_simulated_returns_length");
}

// ---------------------------------------------------------------------------
// Fixture validation
// ---------------------------------------------------------------------------

void test_historical_fixture(const std::vector<double>& returns) {
    auto expected = load_expected(fixtures_dir() + "/expected_historical.csv");
    for (const auto& row : expected) {
        double v = var::var_historical(returns, row.confidence, row.holding_period);
        double cv = var::cvar_historical(returns, row.confidence, row.holding_period);
        std::string label = "hist conf=" + std::to_string(row.confidence)
                          + " hp=" + std::to_string(row.holding_period);
        assert_approx(v, row.var, 0.01, label + " var");
        assert_approx(cv, row.cvar, 0.01, label + " cvar");
    }
    pass("historical_fixture_validation");
}

void test_parametric_fixture(const std::vector<double>& returns) {
    auto expected = load_expected(fixtures_dir() + "/expected_parametric.csv");
    for (const auto& row : expected) {
        double v = var::var_parametric(returns, row.confidence, row.holding_period);
        double cv = var::cvar_parametric(returns, row.confidence, row.holding_period);
        std::string label = "param conf=" + std::to_string(row.confidence)
                          + " hp=" + std::to_string(row.holding_period);
        assert_approx(v, row.var, 0.001, label + " var");
        assert_approx(cv, row.cvar, 0.001, label + " cvar");
    }
    pass("parametric_fixture_validation");
}

// ---------------------------------------------------------------------------
// main
// ---------------------------------------------------------------------------

int main() {
    std::cout << "Loading fixture data..." << std::endl;
    auto returns = load_returns(fixtures_dir() + "/sample_returns.csv");
    std::cout << "Loaded " << returns.size() << " returns\n" << std::endl;

    test_log_returns();

    test_historical_positive(returns);
    test_historical_monotonic_confidence(returns);
    test_historical_monotonic_horizon(returns);
    test_historical_cvar_geq_var(returns);

    test_parametric_positive(returns);
    test_parametric_monotonic_confidence(returns);
    test_parametric_cvar_geq_var(returns);

    test_monte_carlo_positive(returns);
    test_monte_carlo_cvar_geq_var(returns);
    test_monte_carlo_seed_reproducibility(returns);
    test_monte_carlo_simulated_returns_length(returns);

    test_historical_fixture(returns);
    test_parametric_fixture(returns);

    std::cout << "\n" << tests_passed << "/" << tests_run << " tests passed" << std::endl;
    return (tests_passed == tests_run) ? 0 : 1;
}
